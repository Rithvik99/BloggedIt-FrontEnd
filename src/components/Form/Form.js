import React, { useState, useEffect } from 'react';
import { TextField, Button, Typography, Paper ,FormHelperText} from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import FileBase from 'react-file-base64';
import { useHistory } from 'react-router-dom';
import ChipInput from 'material-ui-chip-input';

import { createBlog, updateBlog } from '../../actions/blogs';
import useStyles from './styles';

const Form = ({ currentId, setCurrentId , setDisplayForm}) => {
  const [blogData, setBlogData] = useState({ title: '', message: '', tags: [], selectedFile: '' });
  const blog = useSelector((state) => (currentId ? state.blogs.blogs.find((message) => message._id === currentId) : null));
  const dispatch = useDispatch();
  const classes = useStyles();
  const user = JSON.parse(localStorage.getItem('profile'));
  const history = useHistory();
  const [submitEnabled, setSubmitEnabled] = useState(false);

  const clear = () => {
    setCurrentId(0);
    setBlogData({ title: '', message: '', tags: [], selectedFile: '' });
    setSubmitEnabled(false);
  };
  
  const goBack= () => {
    setDisplayForm(false);
    clear();
  }

  useEffect(() => {
    if (!blog?.title) clear();
    if (blog) setBlogData(blog);
  }, [blog]);

  useEffect(() => {
    setSubmitEnabled(blogData.title.trim() !== '' && blogData.message.trim() !== '');
  }, [blogData]);

  const handleSubmit = async (e) => {
    e.preventDefault();
     
    if (!submitEnabled) return;
    if (currentId === 0) {
      setDisplayForm(false)
      dispatch(createBlog({ ...blogData, name: user?.result?.name }, history));
      clear();
    } else {
      setDisplayForm(false)
      dispatch(updateBlog(currentId, { ...blogData, name: user?.result?.name }));
      clear();
    }
  };

  if (!user?.result?.name) {
    return (
      <Paper className={classes.paper} elevation={6}>
        <Typography variant="h6" align="center">
          Please Sign In to create your own blogs and like other's blogs.
        </Typography>
      </Paper>
    );
  }

  const handleAddChip = (tag) => {
    setBlogData({ ...blogData, tags: [...blogData.tags, tag] });
  };

  const handleDeleteChip = (chipToDelete) => {
    setBlogData({ ...blogData, tags: blogData.tags.filter((tag) => tag !== chipToDelete) });
  };

  return (
    <Paper className={classes.paper} elevation={6}>
      <div className={classes.backArrow} onClick={() => goBack()}>
        &larr; 
      </div>
      <form autoComplete="off" noValidate className={`${classes.root} ${classes.form}`} onSubmit={handleSubmit}>
        <Typography variant="h6">{currentId ? `Editing "${blog?.title}"` : 'Create a Blog'}</Typography>
        <TextField name="title" variant="outlined" label="Title" fullWidth value={blogData.title} onChange={(e) => setBlogData({ ...blogData, title: e.target.value })} />
        <TextField name="message" variant="outlined" label="Message" fullWidth multiline rows={4} value={blogData.message} onChange={(e) => setBlogData({ ...blogData, message: e.target.value })} />
        <div style={{ padding: '5px 0', width: '94%' }}>
          <ChipInput
            name="tags"
            variant="outlined"
            label="Tags"
            fullWidth
            value={blogData.tags}
            onAdd={(chip) => handleAddChip(chip)}
            onDelete={(chip) => handleDeleteChip(chip)}
          />
          <FormHelperText className={classes.helptext}>Press Enter after entering your tag</FormHelperText>
        </div>
        <div className={classes.fileInput}><FileBase type="file" multiple={false} onDone={({ base64 }) => setBlogData({ ...blogData, selectedFile: base64 })} /></div>
        <Button className={classes.buttonSubmit} variant="contained" color="primary" size="large" type="submit" fullWidth disabled={!submitEnabled}>Submit</Button>
        <Button variant="contained" color="secondary" size="small" onClick={clear} fullWidth>Clear</Button>
      </form>
    </Paper>
  );
};

export default Form;