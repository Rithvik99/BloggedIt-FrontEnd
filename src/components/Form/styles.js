
import { makeStyles } from '@material-ui/core/styles';

export default makeStyles((theme) => ({
  root: {
    '& .MuiTextField-root': {
      margin: theme.spacing(1),
    },
  },
  paper: {
    padding: theme.spacing(2),
  },
  form: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'center',
  },
  backArrow: {
    cursor: 'pointer',
    // marginBottom: theme.spacing(1), 
    fontSize: '2rem', 
  },
  fileInput: {
    width: '97%',
    margin: '10px ',
  },
  helptext: {
    marginBottom: theme.spacing(2),
  },
  buttonSubmit: {
    marginBottom: 10,
  },
}));