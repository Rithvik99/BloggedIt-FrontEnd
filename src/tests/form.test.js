/** @jest-environment jsdom */
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import React from 'react';
import Form from '../components/Form/Form'; // Update the import path accordingly

// Mock dependencies
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

describe('Form Component', () => {
  // Mock the Redux store state and actions as needed
  const mockDispatch = jest.fn();
  const mockSelector = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    jest.mock('react-redux', () => ({
      ...jest.requireActual('react-redux'),
      useSelector: jest.fn(),
      useDispatch: () => mockDispatch,
    }));
  });

  
  test('renders form elements', () => {
    const { container: view, getByRole } = render(<Form currentId={0} setCurrentId={() => {}} setDisplayForm={() => {}} />);
    expect(view).toMatchSnapshot();
  });
});