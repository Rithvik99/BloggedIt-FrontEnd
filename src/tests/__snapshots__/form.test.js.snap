// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component renders form elements 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-paper-2 MuiPaper-elevation6 MuiPaper-rounded"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter"
    >
      Please Sign In to create your own blogs and like other's blogs.
    </h6>
  </div>
</div>
`;
